Ilariae/executor (#658)
wormhole-foundation/wormhole-docs: products/messaging/concepts/executor-overview.md

---
title: Executor Overview
description: Learn about the Executor framework - a shared, permissionless system for executing cross-chain messages using standardized contracts and quotes.
categories: Basics
---

# Executor

> The Executor is a shared execution framework used to deliver Wormhole messages across chains. It standardizes how message execution is requested, quoted, and performed, enabling any service or protocol to execute messages permissionlessly through on-chain contracts.

Reviewer: ```suggestion
The Executor is a shared execution framework that delivers Wormhole messages across chains. It standardizes how message execution is requested, quoted, and performed, enabling any service or protocol to execute messages permissionlessly through on-chain contracts.
```

The [Executor framework](https://github.com/wormholelabs-xyz/example-messaging-executor/tree/main){target=\_blank} enables anyone to act as a relayer within a permissionless network that uses a request-and-quote model for delivering messages. Instead of relying on a single, centralized relayer service, the Executor framework creates an open marketplace where multiple providers can compete to deliver messages based on signed execution quotes.

At its core, the Executor relies on Wormhole’s existing guarantees: messages are still secured by VAAs and verified by the Guardian network. The difference lies in how delivery requests are initiated and fulfilled.  

1. Applications call a lightweight, stateless Executor contract on the source chain, providing the target chain, target address, and a signed fee quote from a chosen provider.  
2. The contract emits an event representing the execution request, which any off-chain provider can detect.  
3. A matching provider then retrieves the VAA and performs the delivery on the destination chain.

> By decentralizing message execution and supporting both EVM and non-EVM environments, the Executor framework allows developers to integrate Wormhole relaying with broader chain compatibility and without deploying or maintaining their own relayers.

Reviewer: ```suggestion
By decentralizing message execution and supporting both EVM and non-EVM environments, the Executor framework enables developers to integrate Wormhole relaying with broader chain compatibility, without deploying or maintaining their own relayers.
```

## Components 

- **Relay Provider**: An off-chain party responsible for performing message execution between chains. 

> - **Executor Contract**: The shared on-chain contract or program used to make execution requests. 

Reviewer 2: Link this to the executor contract addresses page

- **Execution Quote**: A signed quote defining cost and parameters for execution between a source and destination chain. 

> - **Execution Request**: A request generated on-chain or off-chain for a given message (e.g., NTT, VAA v1, etc) to be executed on another chain. 

Reviewer 2: ```suggestion
- **Execution Request**: A request generated on-chain or off-chain for a given message (e.g., NTT, VAA v1, etc.) to be executed on another chain. 
```

4. The integrator publishes the message via the Wormhole Core contract.  
5. The integrator then calls the Executor contract to register the execution request.

```mermaid
---
title: v1 VAA Execution Request
---
sequenceDiagram
		participant C as Client
		participant Q as Quoter
		box Source Chain
		participant I as Integrator Contract
		participant W as Wormhole Core
		participant E as Executor Contract
		end
    C->>Q: srcChain, dstChain
    Q-->>C: signedQuote
    C->>I: sendMessage(signedQuote, relayInstructions)
    I->>W: publishMessage
    W-->>I: sequence
    I->>E: requestExecution
```

## Result Flow

Once the request is recorded on-chain, off-chain Relay Providers monitor the Executor contract for events that match their signed quotes. When a valid request is detected, the provider retrieves the message from the Guardians and executes it on the destination chain.

1. The Executor contract emits an event with the request and payment details.
2. A Relay Provider verifies the quote and fetches the associated message (e.g., a VAA).
3. The provider delivers the message to the destination chain’s integrator contract.

> 4. The integrator verifies the message with the Wormhole Core contract and performs the specified logic.

Reviewer 2: ```suggestion
4. The integrator verifies the message with the [Wormhole Core contract](/docs/products/reference/contract-addresses/#core-contracts) and performs the specified logic.
```

```mermaid
---
title: v1 VAA Execution Result
---
sequenceDiagram
		box Source Chain
		participant EC as Executor Contract
		end
		participant E as Relayer (Off-Chain)
		box Destination Chain
		participant I as Integrator Contract
		participant W as Wormhole Core
		end
		EC-->>E: event
    E->>I: executeVaaV1
    I->>W: parseAndVerifyVM
```

## Security Considerations

> The Executor Contract is explicitly designed to be immutable and sit outside an integrator's security stack. Executor is intended to be used as a mechanism to permissionlessly deliver cross-chain data that includes an independent attestation source, such as Wormhole v1 VAAs.

Reviewer 2: ```suggestion
The Executor Contract is explicitly designed to be immutable and sit outside an integrator's security stack. Executor is intended to be used as a mechanism to permissionlessly deliver cross-chain data that includes an independent attestation source, such as Wormhole VAAs.
```

Reviewer 2: We don't have a distinction of different kinds of VAAs in the docs right now. Better to avoid confusion IMO.
